### PR TITLE
WIP(DO NOT MERGE): [sw] Add devicetree abstractions and APIs for multi-top support

### PR DESCRIPTION
**Pull request by a-will**
This PR is primarily for discussion. It shows how devicetrees and Zephyr's APIs could be used to provide an abstraction layer for querying top-specific details about the hardware.

_Note: Even though this PR would make available methods for sharing test and testutils code, it does not mean top-specific tests or top_earlgrey.h would go away. The goal is to enable sharing a broad subset of tests, and this round focuses on base addresses, interrupts, and pinmux mappings._

The commits are grouped by functional addition, in case that helps with viewing. The last one shows ways the APIs could work in (though we might want some a more targeted macro layer to reduce verbosity).

Here's roughly how the flow works:
- topgen would generate a device tree file
- For each platform or execution environment, a developer would manually add the top-level devicetree, which would `#include` the base tree from topgen.
- cpp (the c preprocessor) would resolve any C-style includes and perform macro substitution, to spit out a single DTS (device tree source) file
- Zephyr's `scripts/dts/gen_defines.py` script would run against the pre-processed DTS and produce a `devicetree_generated.h`
    - `devicetree_generated.h` would become part of the standard includes for a test execution environment
- Tests and testutils can then use Zephyr's Generic [DT APIs](https://docs.zephyrproject.org/latest/build/dts/api/api.html) (which are only macros and #include <devicetree_generated.h>)
The "calling" code would #include <zephyr/devicetree.h>

Currently, the modeled flow would treat each execution environment (CW310 vs. CW340 vs. Verilator vs. DV) as being a completely separate platform, since they'd each have their own pin map. That could potentially go a different route, though! Not everything has been figured out yet.

### CODE REVIEW

**Comment by a-will** [comment]
Slotted the device tree headers into execution environment's libraries. (Made `lib` into the plural `libs` for the `exec_env` definition to accommodate.)

Now some targets don't need to explicitly depend on a top--The execution environment machinery will take care of providing the correct header.

File: hw/top_earlgrey/BUILD [hunk 1]
```diff
@@ -120,7 +123,10 @@ fpga_cw305(
     name = "fpga_cw305",
     design = "earlgrey",
     exec_env = "fpga_cw305",
-    lib = "//sw/device/lib/arch:fpga_cw305",
+    libs = [
+        "//sw/device/lib/arch:fpga_cw305",
+        "//hw/top_earlgrey/sw/dts:cw310_devicetree",
```
Comment (a-will) [inline hw/top_earlgrey/BUILD:128]: These of course need to change for this board. I'm just showing the flow for now.

File: sw/device/lib/devicetree/dt.c
Comment (a-will) [inline sw/device/lib/devicetree/dt.c:1]: This file can be ignored right now. It was another possible way to organize the data, but it seems better to have IP-specific files.

**Comment by a-will** [comment]
Looks like LLD is broken and stops the ROM build. It claims that the rom memory is overflowing, but the map and section tables say otherwise:
```
Sections:
Idx Name          Size      VMA       LMA       File off  Algn
  0 .vectors      00000180  00008000  00008000  00001000  2**8
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  1 .crt          00000680  00008180  00008180  00001180  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  2 .text         00005e30  00008800  00008800  00001800  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  3 .shutdown     0000031c  0000e630  0000e630  00007630  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  4 .rodata       000012cc  0000e960  0000e960  00007960  2**5
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  5 .static_critical 00001fe8  10000000  10000000  00009000  2**2
                  ALLOC
  6 .data         00000000  10001fe8  10001fe8  00009000  2**0
                  ALLOC
  7 .bss          000004a0  10001fe8  10001fe8  00009000  2**3
                  ALLOC
  8 .chip_info    0000000c  0000ff80  0000ff80  00009f80  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  9 .logs.fields  00000000  00000000  00000000  00009f8c  2**0
                  CONTENTS, READONLY
 10 .riscv.attributes 00000042  00000000  00000000  000474a5  2**0
                  CONTENTS, READONLY
 11 .ot.status_create_record 00000000  00000000  00000000  000474e7  2**0
                  CONTENTS, READONLY
```

**Comment by pamaury** [comment]
> Looks like LLD is broken and stops the ROM build. It claims that the rom memory is overflowing, but the map and section tables say otherwise:

I think @nbdd0121 might have run into [this issue](https://github.com/lowRISC/opentitan/pull/18765#issuecomment-1582762327) where LLD checks the size before relaxation so it might actually fit but it errors out too early.

**Comment by a-will** [comment]
It seems like this would involve too much interaction with the C preprocessor, which is a cumbersome tool. Retry with auto-generated code in #20320

**Pull request closed without merging**